zoom_monitor.py

```diff
@@ -150,92 +172,74 @@ def _open_participants_list(self):
+                By.CLASS_NAME, "participants-wrapper__inner")
+            self._logger.info("participants list opened")
+            return  # Success!
 
         # If we get here, none of our attempts opened the participants list.
         raise ElementClickInterceptedException(
             f"Could not open participants list after {attempt + 1} attempts")
 
-    def _wait_for_element(self, approach, value):  # Helper function
-        """
-        Wait until there is at least one element identified by the approach
-        and value. If 5 seconds elapse without such an element appearing, we
-        raise an exception.
-        Return the element the first element that is found.
+    def _is_selected(self):
         """
-        WebDriverWait(self._driver, 5).until(lambda _:
-            len(self._driver.find_elements(approach, value)) != 0)
-        return self._driver.find_elements(approach, value)[0]
+        Find the participants icon using the class name.
 
-    def _checkIfMeetingEnded(self):
+        The two classes the participant icon can have are:
+        "SvgParticipantsDefault" - the default button class.
+        "SvgParticipantsHovered" - the button is already selected.
+
+        Return if the participants button is selected or not.
         """
-        Throw a MeetingEndedException if the meeting has been ended by the
-        host, and otherwise do nothing.
+
+        element = self._wait_for_element(By.XPATH, PARTICIPANTS_BTN)
+        return element.get_attribute("class") == "SvgParticipantsHovered"
+
+    def _find_participants_button(self):
         """
-        try:
-            modal_title = self._driver.find_element(
-                By.CLASS_NAME, "zm-modal-body-title")
-        except NoSuchElementException:
-            return
+        We want to click on an item with the participants icon. However, the
+        icon itself is not clickable. A click would be intercepted by its
+        grandparent element, a button with the class
+        "footer-button-base__button". Since it's not obvious how to click an
+        SVG element's grandparent, look through all footer buttons.
 
-        if modal_title.text == "This meeting has been ended by host":
-            self._meeting_ended = True  # Don't try logging out later
-            raise MeetingEndedException()
+        Return the button that contains the participants icon.
+        """
+        for outer in self._driver.find_elements(
+            By.CLASS_NAME, "footer-button-base__button"):
+
+            try:
+                self._logger.debug(
+                    f"trying to find participants default in {outer}")
+                # Check if this footer button contains the participants
+                outer.find_element(By.XPATH, PARTICIPANTS_BTN)
+                return outer
+            except NoSuchElementException:
+                self._logger.debug("participants not present, next...")
+                continue  # wrong footer element, try the next one
 
```

Comment:
If we get to here, we will return `None`, and then calling `.click()` on line 187 will raise an `AttributeError` because the function doesn't exist. Catching the `AttributeError` is usually a bad idea: if you have a typo elsewhere in the program, we'll catch and silently suppress that problem too. I'd either check if the button is `None` on line 180ish, or else raise some kind of `ButtonMissingError` here that gets caught around line 180ish.

Reply:
Ooo good point! Will make these changes~ I think checking if the button is None should be enough!

Reply:
Maybe it would be better to raise a `NoSuchElementException` right here `if outer is None:`?

Reply:
Yeah, I'd be okay with that!